### PR TITLE
Anomaly and Panel Chart data download.

### PR DESCRIPTION
**Pull request by bhargavsk1077**
New API Endpoints to allow download of Anomaly and Panel Chart data in the form of CSV.

The following are the endpoints:
- `<base-url>/api/anomaly-data/<int:kpi_id>/download_anomaly_data`, method - `GET`
- `<base-url>/api/rca/<int:kpi_id>/download_chart_data`, method - `GET`

### CODE REVIEW

File: chaos_genius/core/rca/rca_utils/api_utils.py [hunk 1]
```diff
@@ -116,10 +116,15 @@ def kpi_line_data(kpi_id):
...
-            )
+        if not download:
+            for row in final_data:
+                row["date"] = convert_datetime_to_timestamp(
+                    get_rca_date_from_string(row["date"])
+                )
+        else:
+            for row in final_data:
+                row["date"] = get_rca_date_from_string(row["date"])
+
```
Comment (Samyak2) [inline chaos_genius/core/rca/rca_utils/api_utils.py:127]: Why is this converted only when downloading?

File: chaos_genius/views/anomaly_data_view.py [hunk 2]
```diff
@@ -1,9 +1,14 @@
 # -*- coding: utf-8 -*-
 """anomaly data view."""
 import time
+import io
+import csv
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:5]: These imports are not sorted correctly

File: chaos_genius/views/anomaly_data_view.py [hunk 3]
```diff
@@ -1,9 +1,14 @@
...
 from datetime import date, datetime, timedelta
 from typing import Any, Dict, List, Optional, Tuple, cast
 
+from flask import Blueprint, current_app, jsonify, request, send_file
+from sqlalchemy import func, delete
+
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:11]: These too

File: chaos_genius/views/anomaly_data_view.py [hunk 4]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
     else:
         return jsonify({"msg" : f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
+@blueprint.route("/<int:kpi_id>/download_anomaly_data", methods=["GET"])
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:398]: (flake8) E302 expected 2 blank lines, found 1
```suggestion
        return jsonify({"msg" : f"retraining failed for KPI: {kpi_id}, KPI id is None"})


@blueprint.route("/<int:kpi_id>/download_anomaly_data", methods=["GET"])
```

File: chaos_genius/views/anomaly_data_view.py [hunk 5]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
...
+            raise Exception(f"No anomaly data found for KPI id {kpi_id}")
+        output_csv_obj = io.StringIO()
+        csv_headers = ["datetime",
+                       "value",
+                       "severity",
+                       "upper_bound",
+                       "lower_bound",
+                       "series_type",
+                       "series_name",
+                      ]
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:416]: (flake8) E124 closing bracket does not match visual indentation

Should be formatted as:
```suggestion
        csv_headers = [
            "datetime",
            "value",
            "severity",
            "upper_bound",
            "lower_bound",
            "series_type",
            "series_name",
        ]
```

File: chaos_genius/views/anomaly_data_view.py [hunk 6]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
...
+        csvwriter.writerow(csv_headers)
+        for row in data_points:
+            attr_list = [row.data_datetime.strftime("%a %-d %B %H:%M:%S %Y"),
+                         str(row.y),
+                         str(row.severity),
+                         str(row.yhat_upper),
+                         str(row.yhat_lower),
+                         str(row.anomaly_type),
+                         str(row.series_type),
+                        ]
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:427]: (flake8) E124 closing bracket does not match visual indentation

Should be formatted as:
```suggestion
            attr_list = [
                row.data_datetime.strftime("%a %-d %B %H:%M:%S %Y"),
                str(row.y),
                str(row.severity),
                str(row.yhat_upper),
                str(row.yhat_lower),
                str(row.anomaly_type),
                str(row.series_type),
            ]
```

File: chaos_genius/views/anomaly_data_view.py [hunk 7]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
...
+
+        output_csv_obj.seek(0)
+        output_csv_str = output_csv_obj.read().encode('utf-8')
+        output_csv_bytes = io.BytesIO(output_csv_str)
+        output_csv_obj.close()
+
+        return send_file(output_csv_bytes,
+                            mimetype='text/csv',
+                            attachment_filename=f'KPI-{kpi_id}-anomaly-data.csv',
+                            as_attachment=True)
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:438]: (flake8) E127 continuation line over-indented for visual indent

Should be formatted as:
```suggestion
        return send_file(
            output_csv_bytes,
            mimetype="text/csv",
            attachment_filename=f"KPI-{kpi_id}-anomaly-data.csv",
            as_attachment=True,
        )
```

File: chaos_genius/views/anomaly_data_view.py [hunk 8]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
...
+    except Exception as e:
+        return jsonify({'status':'failure',
+                        'message':'Downloading data failed :{}'.format(e)})
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:441]: (flake8) E231 missing whitespace after ':'

Should be formatted as:
```suggestion
        return jsonify(
            {"status": "failure", "message": "Downloading data failed :{}".format(e)}
        )
```

Comment (Samyak2) [reply chaos_genius/views/anomaly_data_view.py:441]: Also, use an f-string here to be consistent with the rest of the codebase
```suggestion
        return jsonify(
            {"status": "failure", "message": f"Downloading data failed: {e}"}
        )
```

File: chaos_genius/views/rca_view.py [hunk 9]
```diff
@@ -2,7 +2,10 @@
 """Endpoints for data retrieval of computed RCAs."""
 import logging
 
-from flask import Blueprint, jsonify, request  # noqa: F401
+from flask import Blueprint, jsonify, request, send_file  # noqa: F401
+
+import io
+import csv
```
Comment (Samyak2) [inline chaos_genius/views/rca_view.py:8]: imports are not sorted

File: chaos_genius/views/rca_view.py [hunk 10]
```diff
@@ -46,6 +49,40 @@ def kpi_get_line_data(kpi_id):
         message = str(err)
     return jsonify({"status": status, "message": message, "data": data})
 
+@blueprint.route("/<int:kpi>/download_chart_data", methods=["GET"])
```
Comment (Samyak2) [inline chaos_genius/views/rca_view.py:52]: (flake8) E302 expected 2 blank lines, found 1

File: chaos_genius/views/rca_view.py [hunk 11]
```diff
@@ -46,6 +49,40 @@ def kpi_get_line_data(kpi_id):
...
+def kpi_download_line_data(kpi):
+    """API endpoint to download chart data for last 60 days."""
+    try:
+        line_data = kpi_line_data(kpi, download=True)
+        if line_data[0]=="error":
```
Comment (Samyak2) [inline chaos_genius/views/rca_view.py:57]: (flake8) E225 missing whitespace around operator

```suggestion
        if line_data[0] == "error":
```

File: chaos_genius/views/rca_view.py [hunk 12]
```diff
@@ -46,6 +49,40 @@ def kpi_get_line_data(kpi_id):
...
+        output_csv_bytes = io.BytesIO(output_csv_str)
+        output_csv_obj.close()
+
+        return send_file(output_csv_bytes,
+                            mimetype='text/csv',
+                            attachment_filename=f'KPI-{kpi}-panel-chart-data.csv',
+                            as_attachment=True)
+    except Exception as e:
+        return jsonify({"status":"failure",
+                        "message":f"chart data download failed : {e}"})
```
Comment (Samyak2) [inline chaos_genius/views/rca_view.py:84]: needs formatting

**Review by Samyak2 — CHANGES_REQUESTED** [review]

File: chaos_genius/views/anomaly_data_view.py [hunk 13]
```diff
@@ -390,6 +395,50 @@ def kpi_anomaly_retraining(kpi_id):
     else:
         return jsonify({"msg" : f"retraining failed for KPI: {kpi_id}, KPI id is None"})
 
+@blueprint.route("/<int:kpi_id>/download_anomaly_data", methods=["GET"])
+def download_anomaly_data(kpi_id):
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:399]: Would be nice to have type annotations here.

File: chaos_genius/views/anomaly_data_view.py [hunk 14]
```diff
@@ -451,6 +500,23 @@ def convert_to_graph_json(
 
     return graph_data
 
+def get_overall_data_points(kpi_id, n=60):
```
Comment (Samyak2) [inline chaos_genius/views/anomaly_data_view.py:503]: Would be nice to have type annotations here.

File: chaos_genius/views/rca_view.py [hunk 15]
```diff
@@ -46,6 +49,40 @@ def kpi_get_line_data(kpi_id):
         message = str(err)
     return jsonify({"status": status, "message": message, "data": data})
 
+@blueprint.route("/<int:kpi>/download_chart_data", methods=["GET"])
+def kpi_download_line_data(kpi):
```
Comment (Samyak2) [inline chaos_genius/views/rca_view.py:53]: Would be nice to have type annotations here.

**Review by Samyak2 — CHANGES_REQUESTED** [review]

File: chaos_genius/core/rca/rca_utils/api_utils.py (hunk 1 above)
Comment (bhargavsk1077) [reply chaos_genius/core/rca/rca_utils/api_utils.py:127]: The timestamp generated by `convert_datetime_to_timestamp` was not usable as creation of a datetime object from that timestamp failed. So just for downloads the result is returned without the use of the said function.

**Review by Samyak2 — APPROVED** [review]

**Pull request merged**